# 11 错误处理，调试和测试
我们经常会看到很多程序员大部分的"编程"时间都花费在检查bug和修复bug上。无论你是在编写修改代码还是重构系统，几乎都是花费大量的时间在进行故障排除和测试，外界都觉得我们程序员是设计师，能够把一个系统从无做到有，是一项很伟大的工作，而且是相当有趣的工作，但事实上我们每天都是徘徊在排错、调试、测试之间。当然如果你有良好的习惯和技术方案来直面这些问题，那么你就有可能将排错时间减到最少，而尽可能的将时间花费在更有价值的事情上。

但是遗憾的是很多程序员不愿意在错误处理、调试和测试能力上下工夫，导致后面应用上线之后查找错误、定位问题花费更多的时间。所以我们在设计应用之前就做好错误处理规划、测试用例等，那么将来修改代码、升级系统都将变得简单。

开发Web应用过程中，错误自然难免，那么如何更好的找到错误原因，解决问题呢？11.1小节将介绍Go语言中如何处理错误，如何设计自己的包、函数的错误处理，11.2小节将介绍如何使用GDB来调试我们的程序，动态运行情况下各种变量信息，运行情况的监控和调试。

11.3小节将对Go语言中的单元测试进行深入的探讨，并示例如何来编写单元测试，Go的单元测试规则规范如何定义，以保证以后升级修改运行相应的测试代码就可以进行最小化的测试。

长期以来，培养良好的调试、测试习惯一直是很多程序员逃避的事情，所以现在你不要再逃避了，就从你现在的项目开发，从学习Go Web开发开始养成良好的习惯。


 
# 11.1 错误处理
Go语言主要的设计准则是：简洁、明白，简洁是指语法和C类似，相当的简单，明白是指任何语句都是很明显的，不含有任何隐含的东西，在错误处理方案的设计中也贯彻了这一思想。我们知道在C语言里面是通过返回-1或者NULL之类的信息来表示错误，但是对于使用者来说，不查看相应的API说明文档，根本搞不清楚这个返回值究竟代表什么意思，比如:返回0是成功，还是失败,而Go定义了一个叫做error的类型，来显式表达错误。在使用时，通过把返回的error变量与nil的比较，来判定操作是否成功。例如`os.Open`函数在打开文件失败时将返回一个不为nil的error变量

	func Open(name string) (file *File, err error)

下面这个例子通过调用`os.Open`打开一个文件，如果出现错误，那么就会调用`log.Fatal`来输出错误信息：

	f, err := os.Open("filename.ext")
	if err != nil {
		log.Fatal(err)
	}

类似于`os.Open`函数，标准包中所有可能出错的API都会返回一个error变量，以方便错误处理，这个小节将详细地介绍error类型的设计，和讨论开发Web应用中如何更好地处理error。
## Error类型
error类型是一个接口类型，这是它的定义：

	type error interface {
		Error() string
	}

error是一个内置的接口类型，我们可以在/builtin/包下面找到相应的定义。而我们在很多内部包里面用到的 error是errors包下面的实现的私有结构errorString

	// errorString is a trivial implementation of error.
	type errorString struct {
		s string
	}

	func (e *errorString) Error() string {
		return e.s
	}
	
你可以通过`errors.New`把一个字符串转化为errorString，以得到一个满足接口error的对象，其内部实现如下：

	// New returns an error that formats as the given text.
	func New(text string) error {
		return &errorString{text}
	}

下面这个例子演示了如何使用`errors.New`:

	func Sqrt(f float64) (float64, error) {
		if f < 0 {
			return 0, errors.New("math: square root of negative number")
		}
		// implementation
	}
	
在下面的例子中，我们在调用Sqrt的时候传递的一个负数，然后就得到了non-nil的error对象，将此对象与nil比较，结果为true，所以fmt.Println(fmt包在处理error时会调用Error方法)被调用，以输出错误，请看下面调用的示例代码：

	f, err := Sqrt(-1)
    if err != nil {
        fmt.Println(err)
    }	

## 自定义Error
通过上面的介绍我们知道error是一个interface，所以在实现自己的包的时候，通过定义实现此接口的结构，我们就可以实现自己的错误定义，请看来自Json包的示例：

	type SyntaxError struct {
		msg    string // 错误描述
		Offset int64  // 错误发生的位置
	}

	func (e *SyntaxError) Error() string { return e.msg }

Offset字段在调用Error的时候不会被打印，但是我们可以通过类型断言获取错误类型，然后可以打印相应的错误信息，请看下面的例子:

	if err := dec.Decode(&val); err != nil {
		if serr, ok := err.(*json.SyntaxError); ok {
			line, col := findLine(f, serr.Offset)
			return fmt.Errorf("%s:%d:%d: %v", f.Name(), line, col, err)
		}
		return err
	}

需要注意的是，函数返回自定义错误时，返回值推荐设置为error类型，而非自定义错误类型，特别需要注意的是不应预声明自定义错误类型的变量。例如：

	func Decode() *SyntaxError { // 错误，将可能导致上层调用者err!=nil的判断永远为true。
        var err *SyntaxError     // 预声明错误变量
        if 出错条件 {
            err = &SyntaxError{}
        }
        return err               // 错误，err永远等于非nil，导致上层调用者err!=nil的判断始终为true
    }
	
原因见 http://golang.org/doc/faq#nil_error

上面例子简单的演示了如何自定义Error类型。但是如果我们还需要更复杂的错误处理呢？此时，我们来参考一下net包采用的方法：

	package net

	type Error interface {
	    error
	    Timeout() bool   // Is the error a timeout?
	    Temporary() bool // Is the error temporary?
	}

在调用的地方，通过类型断言err是不是net.Error,来细化错误的处理，例如下面的例子，如果一个网络发生临时性错误，那么将会sleep 1秒之后重试：

	if nerr, ok := err.(net.Error); ok && nerr.Temporary() {
		time.Sleep(1e9)
		continue
	}
	if err != nil {
		log.Fatal(err)
	}

## 错误处理
Go在错误处理上采用了与C类似的检查返回值的方式，而不是其他多数主流语言采用的异常方式，这造成了代码编写上的一个很大的缺点:错误处理代码的冗余，对于这种情况是我们通过复用检测函数来减少类似的代码。

请看下面这个例子代码：

	func init() {
		http.HandleFunc("/view", viewRecord)
	}

	func viewRecord(w http.ResponseWriter, r *http.Request) {
		c := appengine.NewContext(r)
		key := datastore.NewKey(c, "Record", r.FormValue("id"), 0, nil)
		record := new(Record)
		if err := datastore.Get(c, key, record); err != nil {
			http.Error(w, err.Error(), 500)
			return
		}
		if err := viewTemplate.Execute(w, record); err != nil {
			http.Error(w, err.Error(), 500)
		}
	}

上面的例子中获取数据和模板展示调用时都有检测错误，当有错误发生时，调用了统一的处理函数`http.Error`，返回给客户端500错误码，并显示相应的错误数据。但是当越来越多的HandleFunc加入之后，这样的错误处理逻辑代码就会越来越多，其实我们可以通过自定义路由器来缩减代码(实现的思路可以参考第三章的HTTP详解)。

	type appHandler func(http.ResponseWriter, *http.Request) error

	func (fn appHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
		if err := fn(w, r); err != nil {
			http.Error(w, err.Error(), 500)
		}
	}

上面我们定义了自定义的路由器，然后我们可以通过如下方式来注册函数：

	func init() {
		http.Handle("/view", appHandler(viewRecord))
	}

当请求/view的时候我们的逻辑处理可以变成如下代码，和第一种实现方式相比较已经简单了很多。

	func viewRecord(w http.ResponseWriter, r *http.Request) error {
		c := appengine.NewContext(r)
		key := datastore.NewKey(c, "Record", r.FormValue("id"), 0, nil)
		record := new(Record)
		if err := datastore.Get(c, key, record); err != nil {
			return err
		}
		return viewTemplate.Execute(w, record)
	}

上面的例子错误处理的时候所有的错误返回给用户的都是500错误码，然后打印出来相应的错误代码，其实我们可以把这个错误信息定义的更加友好，调试的时候也方便定位问题，我们可以自定义返回的错误类型：

	type appError struct {
		Error   error
		Message string
		Code    int
	}

这样我们的自定义路由器可以改成如下方式：

	type appHandler func(http.ResponseWriter, *http.Request) *appError

	func (fn appHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
		if e := fn(w, r); e != nil { // e is *appError, not os.Error.
			c := appengine.NewContext(r)
			c.Errorf("%v", e.Error)
			http.Error(w, e.Message, e.Code)
		}
	}

这样修改完自定义错误之后，我们的逻辑处理可以改成如下方式：

	func viewRecord(w http.ResponseWriter, r *http.Request) *appError {
		c := appengine.NewContext(r)
		key := datastore.NewKey(c, "Record", r.FormValue("id"), 0, nil)
		record := new(Record)
		if err := datastore.Get(c, key, record); err != nil {
			return &appError{err, "Record not found", 404}
		}
		if err := viewTemplate.Execute(w, record); err != nil {
			return &appError{err, "Can't display record", 500}
		}
		return nil
	}

如上所示，在我们访问view的时候可以根据不同的情况获取不同的错误码和错误信息，虽然这个和第一个版本的代码量差不多，但是这个显示的错误更加明显，提示的错误信息更加友好，扩展性也比第一个更好。

## 总结
在程序设计中，容错是相当重要的一部分工作，在Go中它是通过错误处理来实现的，error虽然只是一个接口，但是其变化却可以有很多，我们可以根据自己的需求来实现不同的处理，最后介绍的错误处理方案，希望能给大家在如何设计更好Web错误处理方案上带来一点思路。



# 11.2 使用GDB调试
开发程序过程中调试代码是开发者经常要做的一件事情，Go语言不像PHP、Python等动态语言，只要修改不需要编译就可以直接输出，而且可以动态的在运行环境下打印数据。当然Go语言也可以通过Println之类的打印数据来调试，但是每次都需要重新编译，这是一件相当麻烦的事情。我们知道在Python中有pdb/ipdb之类的工具调试，Javascript也有类似工具，这些工具都能够动态的显示变量信息，单步调试等。不过庆幸的是Go也有类似的工具支持：GDB。Go内部已经内置支持了GDB，所以，我们可以通过GDB来进行调试，那么本小节就来介绍一下如何通过GDB来调试Go程序。

## GDB调试简介
GDB是FSF(自由软件基金会)发布的一个强大的类UNIX系统下的程序调试工具。使用GDB可以做如下事情：

1. 启动程序，可以按照开发者的自定义要求运行程序。
2. 可让被调试的程序在开发者设定的调置的断点处停住。（断点可以是条件表达式）
3. 当程序被停住时，可以检查此时程序中所发生的事。
4. 动态的改变当前程序的执行环境。

目前支持调试Go程序的GDB版本必须大于7.1。

编译Go程序的时候需要注意以下几点

1. 传递参数-ldflags "-s"，忽略debug的打印信息
2. 传递-gcflags "-N -l" 参数，这样可以忽略Go内部做的一些优化，聚合变量和函数等优化，这样对于GDB调试来说非常困难，所以在编译的时候加入这两个参数避免这些优化。 

## 常用命令
GDB的一些常用命令如下所示

- list

	简写命令`l`，用来显示源代码，默认显示十行代码，后面可以带上参数显示的具体行，例如：`list 15`，显示十行代码，其中第15行在显示的十行里面的中间，如下所示。

		10	        time.Sleep(2 * time.Second)
		11	        c <- i
		12	    }
		13	    close(c)
		14	}
		15	
		16	func main() {
		17	    msg := "Starting main"
		18	    fmt.Println(msg)
		19	    bus := make(chan int)

	
- break

	简写命令 `b`,用来设置断点，后面跟上参数设置断点的行数，例如`b 10`在第十行设置断点。
	
- delete
	简写命令 `d`,用来删除断点，后面跟上断点设置的序号，这个序号可以通过`info breakpoints`获取相应的设置的断点序号，如下是显示的设置断点序号。

		Num     Type           Disp Enb Address            What
		2       breakpoint     keep y   0x0000000000400dc3 in main.main at /home/xiemengjun/gdb.go:23
		breakpoint already hit 1 time

- backtrace
	
	简写命令 `bt`,用来打印执行的代码过程，如下所示：

		#0  main.main () at /home/xiemengjun/gdb.go:23
		#1  0x000000000040d61e in runtime.main () at /home/xiemengjun/go/src/pkg/runtime/proc.c:244
		#2  0x000000000040d6c1 in schedunlock () at /home/xiemengjun/go/src/pkg/runtime/proc.c:267
		#3  0x0000000000000000 in ?? ()
- info

	info命令用来显示信息，后面有几种参数，我们常用的有如下几种：
		
	- `info locals`

		显示当前执行的程序中的变量值
	- `info breakpoints`

		显示当前设置的断点列表
	- `info goroutines`

		显示当前执行的goroutine列表，如下代码所示,带*的表示当前执行的

			* 1  running runtime.gosched
			* 2  syscall runtime.entersyscall
			  3  waiting runtime.gosched
			  4 runnable runtime.gosched
- print

	简写命令`p`，用来打印变量或者其他信息，后面跟上需要打印的变量名，当然还有一些很有用的函数$len()和$cap()，用来返回当前string、slices或者maps的长度和容量。

- whatis 
	
	用来显示当前变量的类型，后面跟上变量名，例如`whatis msg`,显示如下：

		type = struct string
- next

	简写命令 `n`,用来单步调试，跳到下一步，当有断点之后，可以输入`n`跳转到下一步继续执行
- coutinue

	简称命令 `c`，用来跳出当前断点处，后面可以跟参数N，跳过多少次断点

- set variable

	该命令用来改变运行过程中的变量值，格式如：`set variable <var>=<value>`

## 调试过程
我们通过下面这个代码来演示如何通过GDB来调试Go程序，下面是将要演示的代码：

	package main

	import (
		"fmt"
		"time"
	)

	func counting(c chan<- int) {
		for i := 0; i < 10; i++ {
			time.Sleep(2 * time.Second)
			c <- i
		}
		close(c)
	}

	func main() {
		msg := "Starting main"
		fmt.Println(msg)
		bus := make(chan int)
		msg = "starting a gofunc"
		go counting(bus)
		for count := range bus {
			fmt.Println("count:", count)
		}
	}

编译文件，生成可执行文件gdbfile:

	go build -gcflags "-N -l" gdbfile.go

通过gdb命令启动调试：

	gdb gdbfile
	
启动之后首先看看这个程序是不是可以运行起来，只要输入`run`命令回车后程序就开始运行，程序正常的话可以看到程序输出如下，和我们在命令行直接执行程序输出是一样的：

	(gdb) run
	Starting program: /home/xiemengjun/gdbfile 
	Starting main
	count: 0
	count: 1
	count: 2
	count: 3
	count: 4
	count: 5
	count: 6
	count: 7
	count: 8
	count: 9
	[LWP 2771 exited]
	[Inferior 1 (process 2771) exited normally]	
好了，现在我们已经知道怎么让程序跑起来了，接下来开始给代码设置断点：

	(gdb) b 23
	Breakpoint 1 at 0x400d8d: file /home/xiemengjun/gdbfile.go, line 23.
	(gdb) run
	Starting program: /home/xiemengjun/gdbfile 
	Starting main
	[New LWP 3284]
	[Switching to LWP 3284]

	Breakpoint 1, main.main () at /home/xiemengjun/gdbfile.go:23
	23	        fmt.Println("count:", count)

上面例子`b 23`表示在第23行设置了断点，之后输入`run`开始运行程序。现在程序在前面设置断点的地方停住了，我们需要查看断点相应上下文的源码，输入`list`就可以看到源码显示从当前停止行的前五行开始：

	(gdb) list
	18	    fmt.Println(msg)
	19	    bus := make(chan int)
	20	    msg = "starting a gofunc"
	21	    go counting(bus)
	22	    for count := range bus {
	23	        fmt.Println("count:", count)
	24	    }
	25	}

现在GDB在运行当前的程序的环境中已经保留了一些有用的调试信息，我们只需打印出相应的变量，查看相应变量的类型及值：

	(gdb) info locals
	count = 0
	bus = 0xf840001a50
	(gdb) p count
	$1 = 0
	(gdb) p bus
	$2 = (chan int) 0xf840001a50
	(gdb) whatis bus
	type = chan int

接下来该让程序继续往下执行，请继续看下面的命令

	(gdb) c
	Continuing.
	count: 0
	[New LWP 3303]
	[Switching to LWP 3303]

	Breakpoint 1, main.main () at /home/xiemengjun/gdbfile.go:23
	23 fmt.Println("count:", count)
	(gdb) c
	Continuing.
	count: 1
	[Switching to LWP 3302]

	Breakpoint 1, main.main () at /home/xiemengjun/gdbfile.go:23
	23 fmt.Println("count:", count)

每次输入`c`之后都会执行一次代码，又跳到下一次for循环，继续打印出来相应的信息。

设想目前需要改变上下文相关变量的信息，跳过一些过程，并继续执行下一步，得出修改后想要的结果：

	(gdb) info locals
	count = 2
	bus = 0xf840001a50
	(gdb) set variable count=9
	(gdb) info locals
	count = 9
	bus = 0xf840001a50
	(gdb) c
	Continuing.
	count: 9
	[Switching to LWP 3302]

	Breakpoint 1, main.main () at /home/xiemengjun/gdbfile.go:23
	23 fmt.Println("count:", count)		
	
最后稍微思考一下，前面整个程序运行的过程中到底创建了多少个goroutine，每个goroutine都在做什么：

	(gdb) info goroutines
	* 1 running runtime.gosched
	* 2 syscall runtime.entersyscall 
	3 waiting runtime.gosched 
	4 runnable runtime.gosched
	(gdb) goroutine 1 bt
	#0 0x000000000040e33b in runtime.gosched () at /home/xiemengjun/go/src/pkg/runtime/proc.c:927
	#1 0x0000000000403091 in runtime.chanrecv (c=void, ep=void, selected=void, received=void)
	at /home/xiemengjun/go/src/pkg/runtime/chan.c:327
	#2 0x000000000040316f in runtime.chanrecv2 (t=void, c=void)
	at /home/xiemengjun/go/src/pkg/runtime/chan.c:420
	#3 0x0000000000400d6f in main.main () at /home/xiemengjun/gdbfile.go:22
	#4 0x000000000040d0c7 in runtime.main () at /home/xiemengjun/go/src/pkg/runtime/proc.c:244
	#5 0x000000000040d16a in schedunlock () at /home/xiemengjun/go/src/pkg/runtime/proc.c:267
	#6 0x0000000000000000 in ?? ()

通过查看goroutines的命令我们可以清楚地了解goruntine内部是怎么执行的，每个函数的调用顺序已经明明白白地显示出来了。

## 小结
本小节我们介绍了GDB调试Go程序的一些基本命令，包括`run`、`print`、`info`、`set variable`、`coutinue`、`list`、`break`	等经常用到的调试命令，通过上面的例子演示，我相信读者已经对于通过GDB调试Go程序有了基本的理解，如果你想获取更多的调试技巧请参考官方网站的GDB调试手册，还有GDB官方网站的手册。	
	


# 11.3 Go怎么写测试用例
开发程序其中很重要的一点是测试，我们如何保证代码的质量，如何保证每个函数是可运行，运行结果是正确的，又如何保证写出来的代码性能是好的，我们知道单元测试的重点在于发现程序设计或实现的逻辑错误，使问题及早暴露，便于问题的定位解决，而性能测试的重点在于发现程序设计上的一些问题，让线上的程序能够在高并发的情况下还能保持稳定。本小节将带着这一连串的问题来讲解Go语言中如何来实现单元测试和性能测试。

Go语言中自带有一个轻量级的测试框架`testing`和自带的`go test`命令来实现单元测试和性能测试，`testing`框架和其他语言中的测试框架类似，你可以基于这个框架写针对相应函数的测试用例，也可以基于该框架写相应的压力测试用例，那么接下来让我们一一来看一下怎么写。

## 如何编写测试用例
由于`go test`命令只能在一个相应的目录下执行所有文件，所以我们接下来新建一个项目目录`gotest`,这样我们所有的代码和测试代码都在这个目录下。

接下来我们在该目录下面创建两个文件：gotest.go和gotest_test.go

1. gotest.go:这个文件里面我们是创建了一个包，里面有一个函数实现了除法运算:

		package gotest
		
		import (
			"errors"
		)
		
		func Division(a, b float64) (float64, error) {
			if b == 0 {
				return 0, errors.New("除数不能为0")
			}
		
			return a / b, nil
		}

2. gotest_test.go:这是我们的单元测试文件，但是记住下面的这些原则：

	- 文件名必须是`_test.go`结尾的，这样在执行`go test`的时候才会执行到相应的代码
	- 你必须import `testing`这个包
	- 所有的测试用例函数必须是`Test`开头
	- 测试用例会按照源代码中写的顺序依次执行
	- 测试函数`TestXxx()`的参数是`testing.T`，我们可以使用该类型来记录错误或者是测试状态
	- 测试格式：`func TestXxx (t *testing.T)`,`Xxx`部分可以为任意的字母数字的组合，但是首字母不能是小写字母[a-z]，例如`Testintdiv`是错误的函数名。
	- 函数中通过调用`testing.T`的`Error`, `Errorf`, `FailNow`, `Fatal`, `FatalIf`方法，说明测试不通过，调用`Log`方法用来记录测试的信息。
	
	下面是我们的测试用例的代码：
	
		package gotest
		
		import (
			"testing"
		)
		
		func Test_Division_1(t *testing.T) {
			if i, e := Division(6, 2); i != 3 || e != nil { //try a unit test on function
				t.Error("除法函数测试没通过") // 如果不是如预期的那么就报错
			} else {
				t.Log("第一个测试通过了") //记录一些你期望记录的信息
			}
		}
		
		func Test_Division_2(t *testing.T) {
			t.Error("就是不通过")
		}

	我们在项目目录下面执行`go test`,就会显示如下信息：

		--- FAIL: Test_Division_2 (0.00 seconds)
			gotest_test.go:16: 就是不通过
		FAIL
		exit status 1
		FAIL	gotest	0.013s
	从这个结果显示测试没有通过，因为在第二个测试函数中我们写死了测试不通过的代码`t.Error`，那么我们的第一个函数执行的情况怎么样呢？默认情况下执行`go test`是不会显示测试通过的信息的，我们需要带上参数`go test -v`，这样就会显示如下信息：
	
		=== RUN Test_Division_1
		--- PASS: Test_Division_1 (0.00 seconds)
			gotest_test.go:11: 第一个测试通过了
		=== RUN Test_Division_2
		--- FAIL: Test_Division_2 (0.00 seconds)
			gotest_test.go:16: 就是不通过
		FAIL
		exit status 1
		FAIL	gotest	0.012s
	上面的输出详细的展示了这个测试的过程，我们看到测试函数1`Test_Division_1`测试通过，而测试函数2`Test_Division_2`测试失败了，最后得出结论测试不通过。接下来我们把测试函数2修改成如下代码：
	
		func Test_Division_2(t *testing.T) {
			if _, e := Division(6, 0); e == nil { //try a unit test on function
				t.Error("Division did not work as expected.") // 如果不是如预期的那么就报错
			} else {
				t.Log("one test passed.", e) //记录一些你期望记录的信息
			}
		}	
	然后我们执行`go test -v`，就显示如下信息，测试通过了：
	
		=== RUN Test_Division_1
		--- PASS: Test_Division_1 (0.00 seconds)
			gotest_test.go:11: 第一个测试通过了
		=== RUN Test_Division_2
		--- PASS: Test_Division_2 (0.00 seconds)
			gotest_test.go:20: one test passed. 除数不能为0
		PASS
		ok  	gotest	0.013s

## 如何编写压力测试
压力测试用来检测函数(方法）的性能，和编写单元功能测试的方法类似,此处不再赘述，但需要注意以下几点：

- 压力测试用例必须遵循如下格式，其中XXX可以是任意字母数字的组合，但是首字母不能是小写字母

		func BenchmarkXXX(b *testing.B) { ... }
		
- `go test`不会默认执行压力测试的函数，如果要执行压力测试需要带上参数`-test.bench`，语法:`-test.bench="test_name_regex"`,例如`go test -test.bench=".*"`表示测试全部的压力测试函数
- 在压力测试用例中,请记得在循环体内使用`testing.B.N`,以使测试可以正常的运行
- 文件名也必须以`_test.go`结尾

下面我们新建一个压力测试文件webbench_test.go，代码如下所示：

	package gotest
	
	import (
		"testing"
	)
	
	func Benchmark_Division(b *testing.B) {
		for i := 0; i < b.N; i++ { //use b.N for looping 
			Division(4, 5)
		}
	}
	
	func Benchmark_TimeConsumingFunction(b *testing.B) {
		b.StopTimer() //调用该函数停止压力测试的时间计数
	
		//做一些初始化的工作,例如读取文件数据,数据库连接之类的,
		//这样这些时间不影响我们测试函数本身的性能
	
		b.StartTimer() //重新开始时间
		for i := 0; i < b.N; i++ {
			Division(4, 5)
		}
	}


我们执行命令`go test -file webbench_test.go -test.bench=".*"`，可以看到如下结果：

	PASS
	Benchmark_Division	500000000	         7.76 ns/op
	Benchmark_TimeConsumingFunction	500000000	         7.80 ns/op
	ok  	gotest	9.364s	

上面的结果显示我们没有执行任何`TestXXX`的单元测试函数，显示的结果只执行了压力测试函数，第一条显示了`Benchmark_Division`执行了500000000次，每次的执行平均时间是7.76纳秒，第二条显示了`Benchmark_TimeConsumingFunction`执行了500000000，每次的平均执行时间是7.80纳秒。最后一条显示总共的执行时间。

## 小结
通过上面对单元测试和压力测试的学习，我们可以看到`testing`包很轻量，编写单元测试和压力测试用例非常简单，配合内置的`go test`命令就可以非常方便的进行测试，这样在我们每次修改完代码,执行一下go test就可以简单的完成回归测试了。



# 11.4 小结
本章我们通过三个小节分别介绍了Go语言中如何处理错误，如何设计错误处理，然后第二小节介绍了如何通过GDB来调试程序，通过GDB我们可以单步调试、可以查看变量、修改变量、打印执行过程等，最后我们介绍了如何利用Go语言自带的轻量级框架`testing`来编写单元测试和压力测试，使用`go test`就可以方便的执行这些测试，使得我们将来代码升级修改之后很方便的进行回归测试。这一章也许对于你编写程序逻辑没有任何帮助，但是对于你编写出来的程序代码保持高质量是至关重要的，因为一个好的Web应用必定有良好的错误处理机制(错误提示的友好、可扩展性)、有好的单元测试和压力测试以保证上线之后代码能够保持良好的性能和按预期的运行。